Aggregated Discovery Endpoint

alexzielenski opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Details in KEP: https://github.com/kubernetes/enhancements/issues/3352

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Adds endpoint at /discovery/v1 which unifies all discovery information known to the cluster in one document.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3352-aggregated-discovery
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/cc @apelisse
/cc @seans3 
/cc @kevindelgado 
/cc @Jefftree 

/sig api-machinery
/triage accepted

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -119,9 +120,11 @@ type Config struct {
+	EnableIndex       bool
+	EnableProfiling   bool
+	EnableDiscovery   bool
+	EnableDiscoveryV1 bool
```
I dont like how the new discovery is V1 and the old discovery is unversioned. Suggestions are appreciated. Would feel slightly weird to introduce new endpoint /discovery/v2 for the first version. Suggestions appreciated

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -643,7 +647,8 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 
-		DiscoveryGroupManager: discovery.NewRootAPIsHandler(c.DiscoveryAddresses, c.Serializer),
+		DiscoveryGroupManager:    discovery.NewRootAPIsHandler(c.DiscoveryAddresses, c.Serializer),
+		DiscoveryResourceManager: discoveryv1.NewResourceManager(c.Serializer),
```
Probably not good naming since the naming of these fields suggests they are connected when they are in fact NOT related (other than both servicing discovery documents)

Jefftree replied:
Maybe something like AggregatedDiscoveryManager or AggregatedDiscoveryResourceManager?

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -673,14 +676,20 @@ func (s *GenericAPIServer) installAPIResources(apiPrefix string, apiGroupInfo *A
 		resourceInfos = append(resourceInfos, r...)
-	}
 
-	s.RegisterDestroyFunc(apiGroupInfo.destroyStorage)
```
I think this was removed by accident?

alexzielenski commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -378,6 +386,39 @@ func (s *APIAggregator) PrepareRun() (preparedAPIAggregator, error) {
+
+	s.GenericAPIServer.AddPostStartHookOrDie("apiservice-discovery-initial", func(context genericapiserver.PostStartHookContext) error {
+		// Call synchronous function to setup initial document once server starts
+		return s.discoveryManager.RefreshDocument()
```
This function waits for ALL discovery documents to refresh (not just crd/builtins). I think that is not desired

apelisse commented on api/api-rules/aggregator_violation_exceptions.list:
```diff
@@ -9,6 +9,11 @@ API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,APIVe
 API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,ApplyOptions,DryRun
 API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,CreateOptions,DryRun
 API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,DeleteOptions,DryRun
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,DiscoveryAPIGroup,Versions
```
We should be able to fix these. If the list must be atomic, then mark them as atomic.

alexzielenski replied:
fixed

alexzielenski committed:
update codegen

alexzielenski committed:
move destroy func into appropriate spot

alexzielenski committed:
fix typo

alexzielenski committed:
add /discovery handler

alexzielenski committed:
avoid blocking on dirtyChannel signal

alexzielenski committed:
add more type annotations

alexzielenski committed:
refactor discovery to use fixed number of webworkers when updating document

alexzielenski committed:
take proxyhandler upfront for service

alexzielenski committed:
fix proto bug

alexzielenski committed:
refactor discoverymanager to be discoveryAggregationController

alexzielenski committed:
only refresh local documents as startup condition

alexzielenski committed:
add boilerplate to new files

alexzielenski committed:
typos, new 1.19 gofmt

alexzielenski commented:
/priority important-soon

alexzielenski committed:
refactor handling of proxyHandler to avoid shared reference bugs

noticed that it is possible for proxyHandler if used as a pointer to be changed whenever APIService changes. avoids a potential race whenever apiservice changes where its proxyhandler points

alexzielenski commented:
Still needs a feature flag and tests but otherwise generally ready for review

alexzielenski committed:
update generated code

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
+// /apis.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DiscoveryAPIGroupList struct {
```
for convenience/normalcy, why not structure this like other list types?

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
+
+// APIGroup contains the name, the supported versions, and the preferred version
+// of a group.
+type DiscoveryAPIGroup struct {
```
Can we structure this with ObjectMeta like a "normal" kind?

Jefftree replied:
Do we need ObjectMeta for this kind of API? It's not a resource belonging to a cluster and just serves as an API format for the /discovery endpoint. Things part of the ObjectMeta definition like ResourceVersion, CreationTimestamp, ManagedFields, etc don't really apply to this type.

Our current apis for handling `/apis` and `/apis/<group>/<version>` also only inherit TypeMeta (https://github.com/kubernetes/kubernetes/blob/6986edae7265a1de536a54ae7ba57767455896b9/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1028-L1034)

lavalamp replied:
Which fields would be set? Part of the goal is to keep the resulting page tiny (so that it's OK for everything to be in one document).

deads2k replied:
> Which fields would be set?

ObjectMeta.Name instead of having the bare `name`.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
+	// name is the name of the group.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// versions are the versions supported in this group.
+	// This will be sorted in descending order based on the preferred version
```
@lavalamp do you know of any case in the wild, where production clusters get versions of things installed that are not in the v1, v2beta1, v1alpha3, style?

lavalamp replied:
I vaguely recall one person wanted to do that, but I don't recall any cases of actually doing that.

negz replied:
The Azure Service Operator uses a somewhat novel variant of the typical v1betaN pattern. Pretty sure it will still be compatible here though: https://azure.github.io/azure-service-operator/reference/

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
+	// corresponding group version's discovery document. This will be nil if the group-version
+	// has not been aggregated yet (APIResources will be empty). To maintain consistency across scenarios with multiple
+	// apiservers, this time will be quantized down to the nearest fifteen minutes.
+	// LastContacted *time.Time `json:"lastContacted" protobuf:"bytes,3,opt,name=lastContacted"`
```
what's the plan with this field?

deads2k replied:
We usually avoid heartbeats because they are expensive to maintain.  While this is (currently) a poll-only API and 15 minutes is fairly long, the benefit of having this field versus having the apiservers remove stale entries or warn when something transitions to stale (status.conditions?) seems more advantageous.

lavalamp replied:
Since users are reading a cache, they need some way to evaluate freshness. We also need to be reasonably sure the caches will converge in the face of apiserver restarts etc.

deads2k replied:
> Since users are reading a cache, they need some way to evaluate freshness. We also need to be reasonably sure the caches will converge in the face of apiserver restarts etc.

Won't this expire the etag every 15 minutes if we serialize it?  It seems more kind to have a condition that indicates, "ContactFailed" that is marked after five minutes of being down, before removal.

lavalamp replied:
Yes, that's a good point. We should just have a "Stale" entry. I forget why we were proposing a timestamp, maybe we weren't thinking we'd literally be hashing the values which include this for the etag.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
+	// from the kubectl CLI interface.
+	SingularName string `json:"singularName" protobuf:"bytes,6,opt,name=singularName"`
+	// namespaced indicates if a resource is namespaced or not.
+	Namespaced bool `json:"namespaced" protobuf:"varint,2,opt,name=namespaced"`
```
in keeping with modern standards, can this be an enum for Scope with value Cluster and Namespace?

lavalamp replied:
+1

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
+}
+
+// APIResource specifies the name of a resource and whether it is namespaced.
+type DiscoveryAPIResource struct {
```
will we nest subresources under this element?  If not, what would an example for say deployments/scale look like?

lavalamp replied:
That's a good question, does the current discovery page list them at all?

deads2k replied:
> That's a good question, does the current discovery page list them at all?

I believe they are listed as peers. That's how we determine if something is scalable for instance.  But logically they are nested.

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/etag.go:
```diff
@@ -0,0 +1,128 @@
+				// Only set these headers when a hash is given.
+				w.Header().Set("Cache-Control", "public, immutable")
+
+				// Set the Expires directive to the maximum value of one year from the request,
```
how about one day instead?  It's still a long time, but if have a bug/problem it can still correct on a practical time scale.

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/etag.go:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	if len(hash) > 0 {
+		if providedHash := reqURL.Query().Get("hash"); len(providedHash) > 0 {
```
can I get a link to the rfc for this?  Apologies if it's obvious, but I was expecting an "If-None-Match" header.

Jefftree replied:
cc @apelisse @alexzielenski: I thought we weren't doing cache busting with aggregated discovery. Is this line + some of the cache busting headers below intentional?

alexzielenski replied:
@Jefftree thats correct I this logic about the "hash" query param is unintentional. I have commit locally which removes it incoming some point tonight

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/etag.go:
```diff
@@ -0,0 +1,128 @@
+		}
+	}
+
+	responsewriters.WriteObjectNegotiated(
```
note to self: come back and check all these parameters.

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/handler.go:
```diff
@@ -0,0 +1,265 @@
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	rdm.apiGroups = nil
```
why aren't apiGroupNames cleared too?

alexzielenski replied:
they are now cleared

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/handler.go:
```diff
@@ -0,0 +1,265 @@
+	defer rdm.lock.Unlock()
+
+	rdm.apiGroups = nil
+	rdm.cachedResponse = nil
```
why isn't the cachedResponseETag cleared?

alexzielenski replied:
it is now being cleared

alexzielenski committed:
cleaer apiGroupNames when resetting resourcemanager

alexzielenski committed:
remove inapplicable openapi cache busting code from discovery etag handler

the hash query parameter is part of cache busting API this code was originally part of, and not intended for general e-tag use

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/handler.go:
```diff
@@ -0,0 +1,265 @@
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedResponse = nil
```
should you only do this if a mutation happened?

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/handler.go:
```diff
@@ -0,0 +1,265 @@
+		// Not very efficient, but in practice there are generally not many versions
+		for i := range existing.Versions {
+			if existing.Versions[i].Version == value.Version {
+				existing.Versions[i] = value
```
check if the old and new are equal before replacing?

deads2k commented on staging/src/k8s.io/apiserver/pkg/endpoints/groupversion.go:
```diff
@@ -105,7 +105,7 @@ type APIGroupVersion struct {
 // It is expected that the provided path root prefix will serve all operations. Root MUST NOT end
 // in a slash.
-func (g *APIGroupVersion) InstallREST(container *restful.Container) ([]*storageversion.ResourceInfo, error) {
+func (g *APIGroupVersion) InstallREST(container *restful.Container) ([]metav1.APIResource, []*storageversion.ResourceInfo, error) {
```
note to self: you got this far and had to fix dinner.

alexzielenski committed:
improve readability

loop was written weirdly due to many edits

alexzielenski committed:
fix bug with versions possibly being dropped

due to using value type. discovered when writing tests

alexzielenski committed:
use correct case for etag

even though headers are case insentitive

alexzielenski committed:
add unit tests for handler and fix data race

alexzielenski committed:
reset all cache fields

unify the fields into a struct so the state is consolidated and can be wiped with a single line

alexzielenski committed:
minor cleanup

alexzielenski committed:
add missing feature gate entry

alexzielenski committed:
change signature to http.Handler

so tests can provide a non-proxy handler

Jefftree commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Discovery API Types
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
```
Prepend all the types in the comments with Discovery since we changed the names (eg: DiscoveryAPIGroupList)

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1113,6 +1113,82 @@ type APIResource struct {
 	StorageVersionHash string `json:"storageVersionHash,omitempty" protobuf:"bytes,10,opt,name=storageVersionHash"`
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```
I guess I expected a meta/v2 with these types being there. I think it's very confusing to have two versions of the same thing in the same place just with different names. Also I see you added conversion functionality -- if you structured it as an actual second version, we have machinery to do most of that for you.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/handler.go:
```diff
@@ -0,0 +1,274 @@
+limitations under the License.
+*/
+
+package v1
```
We already have v1. I think we should start at v2.

deads2k commented:
I think we should refactor the API to reflect the lessons learned from our first discovery document. Details like separation of GVR from GVK and the handling of subresources should be first class concepts this time.  Rather than try to determine that new API in the last two days before freeze, I think we should continue shaping the API with the plan of merging it early in 1.26 instead.  Given that the PR was only opened three days ago, I think the timing on this was always very tight.

@lavalamp sound reasonable?

Some thoughts on the API.
1. Rather than rely on client knowledge of defaults for fields like GVK, I'd like to see it always specified
2. subresource should be nested under the resource and always list GVK
3. need some means of describing endpoints like /logs, /exec, /attach
4. I like the idea of shaping the list like other lists and having an objectmeta.name instead of just the bare name.  I'm open to making a bytes to bytes comparison in proto of the options.
5. it seems like we may want status on these to reflect staleness.

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -887,6 +901,9 @@ func installAPI(s *GenericAPIServer, c *Config) {
 		s.Handler.GoRestfulContainer.Add(s.DiscoveryGroupManager.WebService())
 	}
+	if c.EnableAggregatedDiscoveryEndpoint && utilfeature.DefaultFeatureGate.Enabled(features.AggregatedDiscoveryEndpoint) {
+		s.Handler.GoRestfulContainer.Add(s.DiscoveryResourceManager.WebService())
```
Does the openapi created by this webservice get used somewhere?  If not, are we particularly happy with using gorestful?  I don't hate it, but I don't love it either.

Jefftree replied:
The openapi for this discovery endpoint does get populated in our `/openapi/v2` and `/openapi/v3` documents. Is it not recommended to add additional API endpoints with gorestful?

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -167,6 +173,8 @@ func (cfg *Config) Complete() CompletedConfig {
 	// the kube aggregator wires its own discovery mechanism
 	// TODO eventually collapse this by extracting all of the discovery out
 	c.GenericConfig.EnableDiscovery = false
+	c.GenericConfig.EnableAggregatedDiscoveryEndpoint = false
```
can I get comment about why?  if this is to prevent a conflict with the kube-apiserver's discovery/v1, I would expect it to be set in the `createAggregatorConfig` method in pkg/cmd/kube-apiserver

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -377,6 +385,35 @@ func (s *APIAggregator) PrepareRun() (preparedAPIAggregator, error) {
+		}
+
+		// Setup discovery endpoint
+		s.GenericAPIServer.Handler.GoRestfulContainer.Add(s.discoveryAggregationController.WebService())
```
this surprises me a little.  didn't we just set `EnableAggregatedDiscoveryEndpoint=false` above?  Why do we expose thsi webservice?

alexzielenski replied:
Similar to how today's discovery is implemented, that flag disables the /discovery/v1 functionality within the genericapiserver. The aggregator then sets up its own handler at /discovery/v1 to combine the response from every apiserver registered apiservice into one endpoint.

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:
```diff
@@ -0,0 +1,601 @@
+	LocalName string
+}
+
+// Human-readable String representation used for loggs
```
nit logs

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -426,6 +463,10 @@ func (s *APIAggregator) AddAPIService(apiService *v1.APIService) error {
 		}
+		// Forward calls to discovery manager to update discovery document
+		if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AggregatedDiscoveryEndpoint) {
+			s.discoveryAggregationController.AddAPIService(apiService, proxyHandler)
```
is this called on every update?  For example, is it called when status or labels are updated?

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:
```diff
@@ -0,0 +1,601 @@
+				klog.Errorf("failed to create http.Request for /discovery/v1: %v", err)
+				continue
+			}
+			req.Header.Add("Accept", "application/json")
```
not proto?

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:
```diff
@@ -0,0 +1,601 @@
+			}
+			req.Header.Add("Accept", "application/json")
+
+			if updateInfo.etag != "" {
```
repo style is len(updateInfo.etag) != 0

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:
```diff
@@ -0,0 +1,601 @@
+			}
+
+			writer := newInMemoryResponseWriter()
+			handler.ServeHTTP(writer, req)
```
why are we wiring up an in-memory server acting as a proxy instead of issuing a normal "GET" for the discovery doc directly?

Jefftree replied:
This is issuing a normal "GET" for the discovery doc, similar to how the aggregator aggregates OpenAPI (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go#L57-L69)

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:
```diff
@@ -0,0 +1,601 @@
+	waitGroup := sync.WaitGroup{}
+	results := make(chan resultItem, len(servicesToUpdate))
+
+	webworker := func() {
```
This doesn't appear to handle the case where the aggregated apiserver only exposes the old discovery doc.  Since we cannot ensure every apiserver updates at the same time, it seems useful to allow for both.  even with an  every-minute load on the original discovery doc, we could bet on it being less than having every client request it.

seans3 commented:
> I think we should refactor the API to reflect the lessons learned from our first discovery document. Details like separation of GVR from GVK and the handling of subresources should be first class concepts this time. Rather than try to determine that new API in the last two days before freeze, I think we should continue shaping the API with the plan of merging it early in 1.26 instead. Given that the PR was only opened three days ago, I think the timing on this was always very tight.
> 
> Daniel Smith sound reasonable?
> 
> Some thoughts on the API.
> 
> 1. Rather than rely on client knowledge of defaults for fields like GVK, I'd like to see it always specified
> 2. subresource should be nested under the resource and always list GVK
> 3. need some means of describing endpoints like /logs, /exec, /attach
> 4. I like the idea of shaping the list like other lists and having an objectmeta.name instead of just the bare name.  I'm open to making a bytes to bytes comparison in proto of the options.
> 5. it seems like we may want status on these to reflect staleness.

I think we have to consider the implications of these proposed discovery api changes to the clients. Currently, every client is dependent on the `DiscoveryInterface`, which is expressed in `meta/v1` structures (e.g. `metav1.APIResource`, `metav1.APIGroup`, etc.). And the proposed implementation of the client for this new endpoint merely converts the new `DiscoveryAPI*` structs back to the `API*` structs to ensure no users of the discovery client have to change. If we expose an entire new discovery api that substantially differs from the current api, then every user of the `DiscoveryInterface` will have to change.

The initial impetus for this discovery KEP was to fix the **DiscoveryStorm** problem where clients were periodically downloading dozens of discovery documents regardless of document freshness. It sounds like exposing a new discovery api would radically expand the scope of what we are trying to accomplish here. Should we consider a more modest change without a new discovery api, so it does not cascade to making all clients change?

lavalamp commented:
> Rather than rely on client knowledge of defaults for fields like GVK, I'd like to see it always specified

agree

> subresource should be nested under the resource and always list GVK

Unsure about the former, but adding GVK sounds good.

> need some means of describing endpoints like /logs, /exec, /attach

I think those are already in the open api, and we can keep this discovery doc only about things that need restmapping?

> I like the idea of shaping the list like other lists and having an objectmeta.name instead of just the bare name. I'm open to making a bytes to bytes comparison in proto of the options.

If name is all you want, that sounds fine. If you want e.g. RV simulation then we would need to talk about that more.

> it seems like we may want status on these to reflect staleness.

hm, adding status and metadata clauses does seem like it will make JSON much bigger, at least. But I guess so would adding GVKs everywhere. Part of the reasoning to do this is that the document is small enough that it's reasonable to not split it up on into the foreseeable future.

I'm in favor of making easy & obvious improvements, but I don't think we should make them redo the whole feature -- this was more intended to be a bug fix. We can save complicated stuff for a v3?

lavalamp commented:
@deads2k 

I talked with the team, the intent here was actually not to change the API at all, just do the absolute minimum to get everything into a single document and make it safe given that it's coming out of a cache now. @seans3 believes that it's easy to adjust/update clients if we avoid making semantic changes.

I still advise making a v2 for mechanical reasons (to reuse conversion generators etc), but not making semantic changes. Will that work for you? (I don't expect this to make the code freeze deadline either way) The team didn't intend to sign up for semantic API changes.

I think minimizing the diff for clients is somewhat important, they have to change anyway, but it could be easier or harder to adapt depending on how different we make this document.

apelisse commented:
Actually, it looks like we can re-use the `metav1.APIResource` type directly. That also removes any of the conversion code that currently existed.

alexzielenski committed:
remove DiscoveryAPIResource

it was equivalent to the other type, nice

alexzielenski committed:
minor comment fixup

alexzielenski committed:
use genericfeatures

deads2k commented:
> I talked with the team, the intent here was actually not to change the API at all,

This came up in the KEP review here: https://github.com/kubernetes/enhancements/pull/3364#discussion_r897282810 with a specific comment expanded here for convenience:

> [deads2k](https://github.com/deads2k) [on Jun 14](https://github.com/kubernetes/enhancements/pull/3364#discussion_r897282810)
I think we should make our API serialization the type we want. I can understand if internally you have a RESTMapper convert to something else to make code easier to migrate.

Once a new API is released, it will be pinned. Creating a new API with the known flaws of the previous API is a mistake. The consuming code has to adapt either way.

alexzielenski committed:
renamed to /discovery/v2

regen

alexzielenski committed:
basic aggregated apiservice discovery intregreation test

alexzielenski committed:
update vendor

alexzielenski committed:
minor style fixes

DangerOnTheRanger committed:
Add tests for handler. (#2)

alexzielenski committed:
fixup tests

alexzielenski committed:
use protobuf for downloads between aggregator and apiservers

alexzielenski committed:
use a noopResourcceManager when feature is dsiabled

deads2k commented:
> Once a new API is released, it will be pinned. Creating a new API with the known flaws of the previous API is a mistake. The consuming code has to adapt either way.

An example of the modifications that would provide an API that solves these items: https://github.com/kubernetes/kubernetes/pull/111409#issuecomment-1198536637

alexzielenski committed:
add feature gate

alexzielenski committed:
Add new Discovery types (#3)

* Discovery

* Add discovery types

* generated

alexzielenski commented:
closing this in favor of splitting into smaller PR chunks
